File: LeetCode/Explore/TopInterview/Blind75/Matrix/48_rotate_image.cpp
// Reverse on Diagonal and then Reverse Left to Right

// Complexity Analysis
// Let MM be the number of cells in the grid.
// Time complexity : \mathcal{O}(M)O(M). We perform two steps; transposing the matrix, and then reversing each row. 
// Transposing the matrix has a cost of \mathcal{O}(M)O(M) because we're moving the value of each cell once. 
// Reversing each row also has a cost of \mathcal{O}(M)O(M), because again we're moving the value of each cell once.
// Space complexity : \mathcal{O}(1)O(1) because we do not use any other additional data structures.

class Solution {

private:
  void transpose(vector<vector<int> > &matrix) {
    for (int i = 0; i < matrix.size(); i++) { 
      for (int j = i + 1; j < matrix[0].size(); j++) { 
        swap(matrix[i][j], matrix[j][i]); 
      }
    }
  }
  
  void reversal(vector<vector<int> > &matrix) {
    for (int i = 0; i < matrix.size(); i++) {
      reverse(matrix[i].begin(), matrix[i].end());
    }
  }
  
public:
  void rotate(vector<vector<int>>& matrix) {
    transpose(matrix);
    reversal(matrix);
  }
};

File: LeetCode/Explore/TopInterview/Blind75/Matrix/48_rotate_image.py
class Solution:
  
    def transpose(self, matrix: List[List[int]]):
      for i in range(0, len(matrix), 1):
        for j in range(0, i, 1):
          tmp = matrix[i][j]
          matrix[i][j] = matrix[j][i]
          matrix[j][i] = tmp
          
    def reversal(self, matrix: List[List[int]]):
      for i in range(0, len(matrix), 1):
        matrix[i] = matrix[i][::-1]
          
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        self.transpose(matrix)
        self.reversal(matrix)
